### ISSUE TITLE
Performance Degradation with Reduced Batch Size during Testing on ImageNet-R Dataset

### ISSUE THREAD

**Issue by xvrong** [Minhchuyentoancbn/MoE_PromptCL#1]
Your work is great to me. However, when I tested the trained model, I found that the performance decreased significantly as the batch size was reduced.

Specifically, on the ImageNet-R dataset:

| Batch Size | FAA |
|------------|-----|
| 128        | 75.9510 |
| 64         | 75.9623 |
| 32         | 75.2105 |
| 16         | 75.1027 |
| 8          | 74.7188 |
| 4          | 74.3439 |
| 2          | 73.0702 |
| 1          | 61.8474 |

My training script is as follows:

```bash
for seed in 42 40 44
do
python main.py imr_norgaprompt \
--original_model vit_base_patch16_224 \
--model vit_base_patch16_224 \
--batch-size 128 \
--data-path ./local_datasets/ \
--output_dir ./output/NoRGa_imr_vit_covariance_mlp_2_seed$seed \
--seed $seed \
--train_inference_task_only \
--ca_storage_efficient_method covariance \
--lr 0.0005 --ca_lr 0.005 \
--crct_epochs 30 --epochs 20 \
--shuffle True
done

for seed in 42 40 44
do
python main.py imr_norgaprompt \
--model vit_base_patch16_224 \
--original_model vit_base_patch16_224 \
--batch-size 128 \
--data-path ./local_datasets/ --seed $seed \
--ca_lr 0.005 \
--crct_epochs 30 --epochs 150 \
--prompt_momentum 0.01 --length 20 \
--reg 0.5 \
--sched cosine --larger_prompt_lr \
--ca_storage_efficient_method covariance \
--trained_original_model ./output/NoRGa_imr_vit_covariance_mlp_2_seed$seed \
--output_dir ./output/NoRGa_imr_vit_pe_seed$seed \
--shuffle True --reset --gate_act sigmoid
done
```

My testing script is as follows:

```
for batch in 128 64 32 16 8 4 2 1
do
python main.py imr_norgaprompt \
--model vit_base_patch16_224 \
--original_model vit_base_patch16_224 \
--batch-size $batch \
--data-path ./local_datasets/ --seed 42 \
--ca_lr 0.005 \
--crct_epochs 30 --epochs 150 \
--prompt_momentum 0.01 --length 20 \
--reg 0.5 \
--sched cosine --larger_prompt_lr \
--ca_storage_efficient_method covariance \
--trained_original_model ./output/NoRGa_imr_vit_covariance_mlp_2_seed42 \
--output_dir ./output/NoRGa_imr_vit_pe_seed42 \
--shuffle True --reset --gate_act sigmoid \
--eval 
done
```

Why does this phenomenon occur? Could you provide some explanation and assistance? Thank you.

**Comment by Minhchuyentoancbn** [comment]
Thank you for your interest in our work, and I'm glad you find it helpful. The performance drop you’re observing with decreasing batch sizes is likely due to how the code (originally from the HiDe-Prompt implementation) handles task-adaptive sampling and training hyperparameters. In particular the number of instances sampled per step in task-adaptive training may scale with the batch size. When the batch size is smaller, you might effectively be training on fewer examples or with different sampling proportions per iteration, which can degrade performance.
I hope these suggestions help clarify why performance might degrade with smaller batch sizes and how you can address it. Let me know if you have any other questions!

https://github.com/Minhchuyentoancbn/MoE_PromptCL/blob/838d7f0849c088c6f74c165a8f77f6822bcbb5e5/engines/norga_prompt_engine.py#L521

https://github.com/Minhchuyentoancbn/MoE_PromptCL/blob/838d7f0849c088c6f74c165a8f77f6822bcbb5e5/engines/norga_tii_engine.py#L402

**Comment by xvrong** [comment]
Thank you for your response. However, the observed accuracy drop occurs after the model has been trained and when the trained model is loaded for testing. During training, the batch size remains 128, and only the batch size for testing is adjusted. As in my testing script, I believe the model should directly enter the testing function with this line of code, without using any sampling techniques.
https://github.com/Minhchuyentoancbn/MoE_PromptCL/blob/838d7f0849c088c6f74c165a8f77f6822bcbb5e5/trainers/norgaprompt_trainer.py#L92
Additionally, I have noticed the same phenomenon in the HiDe-Prompt model.

**Comment by xvrong** [comment]
When I keep the batch size at 128 and perform inference on individual samples within a batch, the accuracy drops severely, almost to the level when the batch size is 1.

https://github.com/Minhchuyentoancbn/MoE_PromptCL/blob/838d7f0849c088c6f74c165a8f77f6822bcbb5e5/engines/norga_prompt_engine.py#L126
```
-                
-            output = model(input, task_id=task_id, prompt_id=prompt_id)
-            logits = output['logits']
-            promtp_idx = output['prompt_idx']  # tensor B x topk
+            
+            logits = torch.zeros_like(pretrain_logits)
+            promtp_idx = torch.zeros_like(prompt_id)
+            for idx, sample in enumerate(input):
+                sample_output = model(sample[None], task_id=task_id, prompt_id=prompt_id[idx].view(1, 1))
+                logits[idx] = sample_output['logits']
+                promtp_idx[idx] = sample_output['prompt_idx']  
```
